engine/table/src/main/java/io/deephaven/engine/util/file/TrackedFileHandleFactoryWithLookup.java

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+
+package io.deephaven.engine.util.file;
+
+import io.deephaven.configuration.Configuration;
+import io.deephaven.io.logger.Logger;
+import io.deephaven.io.sched.Scheduler;
+import io.deephaven.net.CommBase;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.nio.file.OpenOption;
+import java.util.Iterator;
+import java.util.List;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.HashMap;
+
+
+/**
+ * An extension of {@link TrackedFileHandleFactory} class with ability to lookup file handles based on file path and
+ * invalidate them in case the underlying file has been modified. This class is used for Parquet files where we
+ * invalidate old file handles when overwriting an existing file.
+ */
+public class TrackedFileHandleFactoryWithLookup extends TrackedFileHandleFactory {
+
+    private static volatile TrackedFileHandleFactoryWithLookup instance;
+
+    /**
+     * Mapping from absolute file path to list of handles
+     */
+    private final Map<String, LinkedList<WeakReference<FileHandle>>> fileToHandleMap = new HashMap<>();
```

Comment:
We need this map and its lists to be accessed in a thread-safe manner. Either `synchronized`, or using concurrent data structures.